### PR TITLE
QgsRasterLayerProperies call QgsMapLayerConfigWidget::apply

### LINKED ISSUE #46832: QgsRasterLayerProperies does not call QgsMapLayerConfigWidget::apply
### What is the bug or the crash?

The vector/raster layer options shown in a QgsVectorLayerProperties/QgsRasterLayerProperties dialog can be enhanced with customized `QgsMapLayerConfigWidgets`. 

After a user triggers the "Accept" or "Apply" button, the dialogs apply method is expected to call the custom widget's apply method as well. However, while this is implemented in `QgsVectorLayerProperties `(see code below), it is missing in `QgsRasterLayerProperties`:

````cpp
void QgsVectorLayerProperties::apply(){
 ...
  // apply all plugin dialogs
  const auto constMLayerPropertiesPages = mLayerPropertiesPages;
  for ( QgsMapLayerConfigWidget *page : constMLayerPropertiesPages )
  {
    page->apply();
  }
 ...
}

````

### Steps to reproduce the issue

This is a unit test to replicate the described behaviour.

````python
import qgis  
import pathlib
import typing
from qgis.PyQt.QtWidgets import QWidget
from qgis.PyQt.QtGui import QIcon
from qgis.gui import QgsMapCanvas, QgsMapLayerConfigWidgetFactory, QgsMapLayerConfigWidget, QgsRasterLayerProperties
from qgis.core import QgsMapLayer, QgsProject, QgsRasterLayer
from utilities import unitTestDataPath
from qgis.testing import start_app, unittest

# Convenience instances in case you may need them
# not used in this test
start_app()


class TestQgsRasterLayerProperties(unittest.TestCase):

    def setUp(self):
        QgsProject.instance().removeAllMapLayers()

    def test_custom_widgets(self):
        class MyWidget(QgsMapLayerConfigWidget):
            """
            A custom QgsMapLayerConfigWidget
            """

            COUNT = 0

            def __init__(self, layer: QgsMapLayer, canvas: QgsMapCanvas, parent: QWidget = None):
                super().__init__(layer, canvas, parent=parent)

            def apply(self) -> None:
                MyWidget.COUNT += 1

        class MyFactory(QgsMapLayerConfigWidgetFactory):
            """
            A custom QgsMapLayerConfigWidgetFactory
            """

            COUNT = 0

            def __init__(self, title: str, icon: QIcon):
                super(MyFactory, self).__init__(title, icon)

            def supportsLayer(self, layer):
                return True

            def supportLayerPropertiesDialog(self):
                return True

            def createWidget(self,
                             layer: QgsMapLayer,
                             canvas: QgsMapCanvas,
                             dockWidget: bool = ..., parent:
                    typing.Optional[QWidget] = ...) -> QgsMapLayerConfigWidget:
                MyFactory.COUNT += 1
                w = MyWidget(layer, canvas, parent=parent)
                return w

        myFactory = MyFactory('Dummy Factory', QIcon())
        myCanvas = QgsMapCanvas()

        myPath = pathlib.Path(unitTestDataPath('raster')) / 'band1_float32_noct_epsg4326.tif'

        myRasterLayer = QgsRasterLayer(myPath.as_posix(), myPath.name)

        assert myRasterLayer.isValid(), f'Raster not loaded {myPath}'

        dialog = QgsRasterLayerProperties(myRasterLayer, myCanvas)

        dialog.addPropertiesPageFactory(myFactory)
        dialog.show()

        # this triggers the dialog's apply
        dialog.accept()
        self.assertEqual(MyFactory.COUNT, 1, msg=f'Custom QgsMapLayerConfigWidget::createWidget(...) not called')
        self.assertEqual(MyWidget.COUNT, 1, msg=f'Custom QgsMapLayerConfigWidget::apply() not called')
````

### Versions

3.23.0-Master
code revision 18979e579c

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jakimowb**
## Description

This PR closes #46832  (QgsRasterLayerProperies does not call QgsMapLayerConfigWidget::apply)

Changes:

- `src/gui/raster/qgsrasterlayerproperties.cpp`: added a loop to call QgsMapLayerConfigWidget::apply of custom pages

- created new test in `tests/src/python/test_qgsrasterlayerproperties.py` and added it to `tests/src/python/CMakeLists.txt`

### CODE REVIEW

File: src/gui/raster/qgsrasterlayerproperties.cpp [hunk 1]
```diff
@@ -835,6 +835,14 @@ void QgsRasterLayerProperties::apply()
   if ( !mRasterLayer->isValid() )
     return;
 
+  // apply all plugin dialogs
+  const auto constMLayerPropertiesPages = mLayerPropertiesPages;
```
Comment (elpaso) [inline src/gui/raster/qgsrasterlayerproperties.cpp:839]: Sorry but `auto` is not much appreciated by some core developers, can you use the real type here?

File: src/gui/raster/qgsrasterlayerproperties.cpp [hunk 2]
```diff
@@ -835,6 +835,14 @@ void QgsRasterLayerProperties::apply()
...
+  for ( QgsMapLayerConfigWidget *page : constMLayerPropertiesPages )
```
Comment (elpaso) [inline src/gui/raster/qgsrasterlayerproperties.cpp:840]: Or better:

```suggestion
  for ( QgsMapLayerConfigWidget *page : std::as_const(mLayerPropertiesPages) )
```

Comment (jakimowb) [reply src/gui/raster/qgsrasterlayerproperties.cpp]: Ok, thanks. 
I don't know much about C++ and just copied these lines from qgsvectorlayerproperties.cpp

**Comment by roya0045** [comment]
If you want to link the pr to the issue see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**